src/npem.c

```diff
@@ -244,3 +273,86 @@ char *npem_get_path(const char *cntrl_path)
 {
 	return str_dup(cntrl_path);
 }
+
+status_t npem_get_slot(char *device, char *slot_path, struct slot_response *slot_res)
+{
+	struct pci_dev *pdev = NULL;
+	struct block_device *block_device = NULL;
+	struct pci_access *pacc = NULL;
+	status_t status = STATUS_SUCCESS;
+	char *path = NULL;
+	u32 reg;
+
+	if (device && device[0] != '\0') {
+		block_device = get_block_device_from_sysfs_path(basename(device));
+		if (block_device)
+			path = block_device->cntrl->sysfs_path;
+	}
+	if (slot_path && slot_path[0] != '\0') {
+		struct cntrl_device *ctrl_dev;
+
+		list_for_each(sysfs_get_cntrl_devices(), ctrl_dev) {
+			if (!is_npem_capable(ctrl_dev->sysfs_path))
+				continue;
+			if (strcmp(basename(ctrl_dev->sysfs_path), basename(slot_path)) != 0)
+				continue;
+			path = ctrl_dev->sysfs_path;
+			block_device = get_block_device_from_sysfs_path(path);
+			break;
+		}
+	}
+	if (block_device)
+		snprintf(slot_res->device, PATH_MAX, "/dev/%s", basename(block_device->sysfs_path));
+	else
+		snprintf(slot_res->device, PATH_MAX, "(empty)");
+
+	if (path)
+		pdev = get_pci_dev_by_path(path, pacc);
+	if (pdev == NULL) {
+		log_error("NPEM: Unable to get pci device for %s\n", path);
+		if (pacc)
+			pci_cleanup(pacc);
+		return STATUS_NULL_POINTER;
+	}
+
+	reg = read_npem_register(pdev, PCI_NPEM_CTRL_REG);
+	slot_res->state = npem_capability_to_ibpi(reg);
+	snprintf(slot_res->slot, PATH_MAX, "%s", path);
+
+	pci_free_dev(pdev);
+	if (pacc)
+		pci_cleanup(pacc);
+	return status;
+}
+
+status_t npem_set_slot(char *slot_path, enum ibpi_pattern state)
+{
+	struct pci_dev *pdev = NULL;
+	struct pci_access *pacc = NULL;
+	status_t status = STATUS_SUCCESS;
+	u32 val;
+	u32 reg;
+
+	pdev = get_pci_dev_by_path(slot_path, pacc);
```

Comment:
you are losing reference to *pacc. pacc needs to be initialized in npem_set_slot()